applications/GeoMechanicsApplication/custom_constitutive/linear_elastic_law.cpp

```diff
@@ -0,0 +1,202 @@
+        this->CalculatePK2Stress(r_strain_vector, r_stress_vector, rParameterValues);
+
+        rValue = 0.5 * inner_prod(r_strain_vector, r_stress_vector); // Strain energy = 0.5*E:C:E
+    }
+
+    return rValue;
+}
+
+Vector& GeoLinearElasticLaw::CalculateValue(ConstitutiveLaw::Parameters& rParameterValues,
+                                            const Variable<Vector>& rThisVariable,
+                                            Vector& rValue)
+{
+    if (rThisVariable == STRAIN || rThisVariable == GREEN_LAGRANGE_STRAIN_VECTOR ||
+        rThisVariable == ALMANSI_STRAIN_VECTOR)
+    {
+        this->CalculateCauchyGreenStrain(rParameterValues, rValue);
+    }
+    else if (rThisVariable == STRESSES || rThisVariable == CAUCHY_STRESS_VECTOR ||
+             rThisVariable == KIRCHHOFF_STRESS_VECTOR || rThisVariable == PK2_STRESS_VECTOR)
+    {
+        // Get Values to compute the constitutive law:
+        Flags& r_flags = rParameterValues.GetOptions();
+
+        // Previous flags saved
+        const bool flag_strain = r_flags.Is(ConstitutiveLaw::USE_ELEMENT_PROVIDED_STRAIN);
+        const bool flag_const_tensor =
+            r_flags.Is(ConstitutiveLaw::COMPUTE_CONSTITUTIVE_TENSOR);
+        const bool flag_stress = r_flags.Is(ConstitutiveLaw::COMPUTE_STRESS);
+
+        r_flags.Set(ConstitutiveLaw::USE_ELEMENT_PROVIDED_STRAIN, false);
+        r_flags.Set(ConstitutiveLaw::COMPUTE_CONSTITUTIVE_TENSOR, true);
+        r_flags.Set(ConstitutiveLaw::COMPUTE_STRESS, true);
+
+        // We compute the stress
+        CalculateMaterialResponseCauchy(rParameterValues);
+        rValue = rParameterValues.GetStressVector();
+
+        // Previous flags restored
+        r_flags.Set(ConstitutiveLaw::USE_ELEMENT_PROVIDED_STRAIN, flag_strain);
+        r_flags.Set(ConstitutiveLaw::COMPUTE_CONSTITUTIVE_TENSOR, flag_const_tensor);
+        r_flags.Set(ConstitutiveLaw::COMPUTE_STRESS, flag_stress);
+    }
+
+    return rValue;
+}
+
+Matrix& GeoLinearElasticLaw::CalculateValue(ConstitutiveLaw::Parameters& rParameterValues,
+                                            const Variable<Matrix>& rThisVariable,
+                                            Matrix& rValue)
+{
+    if (rThisVariable == CONSTITUTIVE_MATRIX || rThisVariable == CONSTITUTIVE_MATRIX_PK2 ||
+        rThisVariable == CONSTITUTIVE_MATRIX_KIRCHHOFF)
+    {
+        CalculateElasticMatrix(rValue, rParameterValues);
+    }
+
+    return rValue;
+}
+
+void GeoLinearElasticLaw::SetValue(const Variable<double>&, const double&, const ProcessInfo&)
+{
+    // Since `ConstitutiveLaw::SetValue` always throws an exception, an override
+    // is required when that is undesired behavior
+}
+
+void GeoLinearElasticLaw::SetValue(const Variable<Vector>&, const Vector&, const ProcessInfo&)
+{
+    // Since `ConstitutiveLaw::SetValue` always throws an exception, an override
+    // is required when that is undesired behavior
+}
```

Comment:
These functions were not there in e.g. the k0 3d law, why are they added here?

Reply:
Nope, I overlooked them, my bad (missed them in the cpp and didn't check for an implementation in the header)